dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/route/RouteService.java

```
@@ -0,0 +1,160 @@
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+import javax.annotation.Nonnull;
+import javax.servlet.http.HttpServletRequest;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import org.hisp.dhis.common.auth.ApiTokenAuth;
+import org.hisp.dhis.common.auth.Auth;
+import org.hisp.dhis.common.auth.HttpBasicAuth;
+import org.jasypt.encryption.pbe.PBEStringCleanablePasswordEncryptor;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.http.*;
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StreamUtils;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * @author Morten Olav Hansen
+ */
+@Service
+@Slf4j
+@RequiredArgsConstructor
+public class RouteService
+{
+    private final RouteStore routeStore;
+
+    private final ObjectMapper objectMapper;
+
+    @Qualifier( AES_128_STRING_ENCRYPTOR )
+    private final PBEStringCleanablePasswordEncryptor encryptor;
+
+    private static final RestTemplate restTemplate = new RestTemplate();
+
+    static
+    {
+        HttpComponentsClientHttpRequestFactory requestFactory = new HttpComponentsClientHttpRequestFactory();
+        requestFactory.setConnectionRequestTimeout( 1_000 );
+        requestFactory.setConnectTimeout( 5_000 );
+        requestFactory.setReadTimeout( 10_000 );
+        requestFactory.setBufferRequestBody( true );
+
+        restTemplate.setRequestFactory( requestFactory );
+    }
+
+    public Route getDecryptedById( @Nonnull String id )
+    {
+        Route route = routeStore.getByUidNoAcl( id );
+
+        if ( route == null )
+        {
+            return null;
```

Do we want to throw an exception here, i.e. is this an invalid state/programming error or user error? Could silently return null hide problems and make it hard to debug?

---

I don't really like throwing http messages from the core layer, that's why i return `null` == 404, I think if we want that we should do a larger refactor

---

Right, I did not mean HTTP message, rather one of our existing exceptions, but fair enough.